Python subinterpreter issues

benjamin-sch opened the pull request:
This should avoid the crash reported in #2051 

1) the first commit counts the "capsules" used. It may be greater 1 if subinterpreters are used. As the InitializeModule seems to initialize the types only once in the case of multiple subinterpreters the deletion of the types should be as well only being done once. Otherwise we get the negative refcnt exception.
2) the second commit always gets the type_pointer from the "capsule" as it may be different after a re-initialization if SWIG is used with python embedded in an application and the application deinitializes/initializes python more than once. Furthermore it might be different if subinterpreters are used as each subinterpreter will add its own "capsule" in SWIG_SetModule.

ojwb commented on Lib/python/pyrun.swg:
```diff
@@ -1627,28 +1627,33 @@ SWIG_Python_TypeCache(void) {
   swig_module_info *swig_module = (swig_module_info *) PyCapsule_GetPointer(obj, SWIGPY_CAPSULE_NAME);
   swig_type_info **types = swig_module->types;
   size_t i;
+  if (--interpreter_counter != 0) // another sub-interpreter nay be still using the swig_module's types
```
comment typo "nay", should I assume be "may"

benjamin-sch replied:
Thanks. You're right. I've fixed it.

ojwb commented:
This looks plausible to me and the appveyor failures are unrelated, but someone more active in the python side of things should probably review.

It should probably have a `CHANGES.current` entry.

BenjaminSbj commented:
> This looks plausible to me and the appveyor failures are unrelated, but someone more active in the python side of things should probably review.
> 
> It should probably have a `CHANGES.current` entry.

Thanks for the hint. I have added an entry to CHANGES.current

benjamin-sch commented:
> This looks plausible to me and the appveyor failures are unrelated, but someone more active in the python side of things should probably review.
> 
> It should probably have a `CHANGES.current` entry.

Thanks for the hint. I've rebased the PR and added an entry to CHANGES.current

ojwb commented:
@wsfulton You should probably review this...

ojwb commented:
@wsfulton wrote on swig-devel:
> If there are any small patches, please tag them with this milestone for consideration for the release.

Doing so...

wsfulton commented:
I've merged this, please check it as there was a merge conflict I had to resolve and I don't have a way to check this code is working with multiple interpreters. On that note, I've not come across sub-interpreters. Are you using https://docs.python.org/3/c-api/init.html#sub-interpreter-support ? If so, I might add a simple mention of these APIs to your new interpreter_counter code.
